Summarize model size in MegaBytes [WIP]

kartik4949 opened the pull request:
Fixes # (issue)
#4807 
This PR adds a Summarization of the model with estimated total model size.
Estimated Total model size is  "input_size + forward/backward pass size + total params size".

e.g
VGG16 with input (1, 1, 28, 28)
Input Size (MBs) --> 0.57
Total Params Size (MBs) --> 528 (e.g downloading weights size)
Forward/Backward Size (MBs) --> 218
Total Estimated Model Size (MBs) --> 747

Testing
--------------------------------------------------------------------------------------------------------------------------------
1. Test knownet (a simple convolution NN)
    tests knownet model with different input shapes and batch sizes.
    we test the corresponding model sizes with pre calculated model size.
2. Test Different Input Types:
    tests a DummyNetwork with different input types.


## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

<!-- For CHANGELOG separate each item in the unreleased section by a blank line to reduce collisions -->

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

kartik4949 commented:
WIP please suggest how to proceed from here.

kartik4949 commented:
@awaelchli @rohitgr7  I'm sensing a usage of LRU cache in model summary, as calculating out and input sizes can be redundant,
thoughts?

tchaton reviewed (commented):
Would you mind adding some tests ?

kartik4949 commented:
> Would you mind adding some tests ?

sure wont mind adding tests :)

pep8speaks commented:
Hello @kartik4949! Thanks for updating this PR. 

* In the file [`pytorch_lightning/core/memory.py`](https://github.com/PyTorchLightning/pytorch-lightning/blob/c6a0355ca9ef3e3180691dd1ccd44e03a3ef533f/pytorch_lightning/core/memory.py):

> [Line 226:17](https://github.com/PyTorchLightning/pytorch-lightning/blob/c6a0355ca9ef3e3180691dd1ccd44e03a3ef533f/pytorch_lightning/core/memory.py#L226): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare 'except'

* In the file [`tests/core/test_memory.py`](https://github.com/PyTorchLightning/pytorch-lightning/blob/c6a0355ca9ef3e3180691dd1ccd44e03a3ef533f/tests/core/test_memory.py):

> [Line 24:5](https://github.com/PyTorchLightning/pytorch-lightning/blob/c6a0355ca9ef3e3180691dd1ccd44e03a3ef533f/tests/core/test_memory.py#L24): [E731](https://duckduckgo.com/?q=pep8%20E731) do not assign a lambda expression, use a def

justusschock commented on pytorch_lightning/core/memory.py:
```diff
@@ -213,6 +213,56 @@ def out_sizes(self) -> List:
+        return sum(self.param_nums)
+
+    @property
+    def model_size(self, input_size=None) -> float:
```
If you define this as a property, you can never pass an argument here...

kartik4949 replied:
@justusschock  yeah was removing this but missed in commit

kartik4949 commented:
@justusschock can u just tests?

justusschock commented:
You could start with simple models like one or two layers, where you know the size / can calculate them by hand and then compare it to the automatically inferred size. Also you could do it once for a more complex model.

kartik4949 commented:
> You could start with simple models like one or two layers, where you know the size / can calculate them by hand and then compare it to the automatically inferred size. Also you could do it once for a more complex model.

Thanks doing the same:)

kartik4949 commented:
> Fixes # (issue)
> #4807
> This PR adds a Summarization of the model with estimated total model size.
> Estimated Total model size is "input_size + forward/backward pass size + total params size".
> 
> e.g
> VGG16 with input (1, 1, 28, 28)
> Input Size (MBs) --> 0.57
> Total Params Size (MBs) --> 528 (e.g downloading weights size)
> Forward/Backward Size (MBs) --> 218
> Total Estimated Model Size (MBs) --> 747
> 
> ## Testing
> 1. Test knownet (a simple convolution NN)
>    tests knownet model with different input shapes and batch sizes.
>    we test the corresponding model sizes with pre calculated model size.
> 2. Test Different Input Types:
>    tests a DummyNetwork with different input types.
> 
> ## Before submitting
> * [x]  Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
> * [x]  Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
> * [x]  Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
> * [ ]  Did you make sure to update the documentation with your changes?
> * [ ]  Did you write any new necessary tests?
> * [ ]  Did you verify new and existing tests pass locally with your changes?
> * [ ]  If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?
> 
> ## PR review
> Anyone in the community is free to review the PR once the tests have passed.
> Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:
> 
> * [ ]  Is this pull request ready for review? (if not, please submit in draft mode)
> * [ ]  Check that all items from **Before submitting** are resolved
> * [ ]  Make sure the title is self-explanatory and the description concisely explains the PR
> * [ ]  Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
> 
> ## Did you have fun?
> Make sure you had fun coding

@awaelchli @ananyahjha93 @williamFalcon @Borda I think this PR is ready , need to add some complex models for testing
i.e 1. multiple input model, 2. multiple blocks 3. multi output.
then we are good to go! :)

Borda commented on pytorch_lightning/core/memory.py:
```diff
@@ -213,6 +214,54 @@ def out_sizes(self) -> List:
+                try:
+                    # try to find prod, i.e check for unknown sizes.
+                    _total_out_params += np.prod(out_sizes)
+                except:
```
can we better specify the exception...

kartik4949 replied:
sure

Borda commented on pytorch_lightning/core/memory.py:
```diff
@@ -213,6 +214,54 @@ def out_sizes(self) -> List:
+    def total_params(self) -> int:
+        return sum(self.param_nums)
+
+    def model_size(self, input_size=None) -> float:
```
let's add typing for all functions and docstrings...
it is not very clear how the input size works with `model. example_input_array`

kartik4949 replied:
input_size is to check model_size on custom input_size passing directly  to model_size

kartik4949 replied:
@Borda im droping the idea of custom input size calculation.

Borda commented on pytorch_lightning/core/memory.py:
```diff
@@ -213,6 +214,54 @@ def out_sizes(self) -> List:
+        return self._get_total_size(in_features if not input_size else input_size)
+
+    def _get_total_size(self, input_size: tuple) -> float:
+        total_input_dsize = abs(np.prod(np.array(input_size))) * self._precision_bytes / (1024 ** 2.0)
```
let's define in init also
```
self._precision_megabytes = self._precision_bytes / (1024 ** 2.0)
```

kartik4949 replied:
@Borda done

Borda reviewed (changes requested):
nice work, just need to finish some minor nits...

kartik4949 commented:
@Borda dropped the idea of input size, we are good with example_input_array only which will act as custom input_size.

kartik4949 commented:
@Borda 
so made new changes,
1. Dropped input_size, example_array will be used instead
2. Model size shown only in FULL mode.
3. Added new test which checks for nested models.

All Tests are passed :)

carmocca commented on tests/core/test_memory.py:
```diff
@@ -81,33 +121,39 @@ def test_empty_model_summary_shapes(mode):
+@pytest.mark.parametrize(
+    ["device"],
+    [
+        pytest.param(torch.device("cpu")),
+        pytest.param(torch.device("cuda", 0)),
+        pytest.param(torch.device("cuda", 0)),
+    ],
+)
 @pytest.mark.skipif(not torch.cuda.is_available(), reason="Test requires GPU.")
```
The `cuda:0` test case here is duplicated, additionally, you want the test to run for the CPU case

```suggestion
        torch.device("cpu"),
        pytest.param(
            torch.device("cuda", 0), marks=pytest.mark.skipif(not torch.cuda.is_available(), reason="Test requires GPU")
        ),
    ],
)
```

carmocca replied:
Yes, I noticed it was there before, but I don't think there is any benefit in testing for both cuda:0 and cuda:1 :thinking:

kartik4949 replied:
Makes Sense!

carmocca commented on tests/core/test_memory.py:
```diff
@@ -119,19 +165,22 @@ def test_mixed_dtype_model_summary():
-    pytest.param(ModelSummary.MODE_FULL),
-    pytest.param(ModelSummary.MODE_TOP),
-])
+@pytest.mark.parametrize(
+    ["mode"],
+    [
+        pytest.param(ModelSummary.MODE_FULL),
+        pytest.param(ModelSummary.MODE_TOP),
+    ],
+)
```
```suggestion
@pytest.mark.parametrize(["mode"], ModelSummary.MODES)
```

carmocca commented on tests/core/test_memory.py:
```diff
@@ -142,10 +191,13 @@ def test_hooks_removed_after_summarize(mode):
-    pytest.param(ModelSummary.MODE_FULL),
-    pytest.param(ModelSummary.MODE_TOP),
-])
+@pytest.mark.parametrize(
+    ["mode"],
+    [
+        pytest.param(ModelSummary.MODE_FULL),
+        pytest.param(ModelSummary.MODE_TOP),
+    ],
+)
```
```suggestion
@pytest.mark.parametrize(["mode"], ModelSummary.MODES)
```

carmocca commented on tests/core/test_memory.py:
```diff
@@ -163,16 +215,16 @@ def test_rnn_summary_shapes(mode):
-    pytest.param(ModelSummary.MODE_FULL),
-    pytest.param(ModelSummary.MODE_TOP),
-])
+@pytest.mark.parametrize(
+    ["mode"],
+    [
+        pytest.param(ModelSummary.MODE_FULL),
+        pytest.param(ModelSummary.MODE_TOP),
+    ],
+)
```
```suggestion
@pytest.mark.parametrize(["mode"], ModelSummary.MODES)
```

carmocca replied:
It works if you do instead:

`@pytest.mark.parametrize("mode", ModelSummary.MODES)`

Same for all other ocurrences.

Sorry for the confusion!

carmocca commented on tests/core/test_memory.py:
```diff
@@ -163,16 +215,16 @@ def test_rnn_summary_shapes(mode):
-        [[b, t, h], [[1, b, h], [1, b, h]]],    # rnn
-        [b, t, o]                               # linear
-    ]
+    assert summary.out_sizes == [[[b, t, h], [[1, b, h], [1, b, h]]], [b, t, o]]  # rnn  # linear
```
This change made the comments not indicate properly what is each part

kartik4949 replied:
formatter :(

carmocca replied:
You can avoid the formatting by adding the "Magic comma" https://github.com/psf/black/issues/1288 (`[b, t, o],`)

kartik4949 replied:
@carmocca thanks, need to setup my new plugin :) will add it in the config.

carmocca commented on tests/core/test_memory.py:
```diff
@@ -186,36 +238,92 @@ def test_summary_parameter_count(mode):
-    pytest.param(ModelSummary.MODE_FULL),
-    pytest.param(ModelSummary.MODE_TOP),
-])
+@pytest.mark.parametrize(
+    ["mode"],
+    [
+        pytest.param(ModelSummary.MODE_FULL),
+        pytest.param(ModelSummary.MODE_TOP),
+    ],
+)
```
```suggestion
@pytest.mark.parametrize(["mode"], ModelSummary.MODES)
```

carmocca commented on tests/core/test_memory.py:
```diff
@@ -186,36 +238,92 @@ def test_summary_parameter_count(mode):
-    pytest.param([torch.zeros(2, 3), torch.zeros(4, 5)], [[2, 3], [4, 5]]),
-    pytest.param((torch.zeros(2, 3), torch.zeros(4, 5)), [[2, 3], [4, 5]]),
-])
+@pytest.mark.parametrize(
+    ["mode"],
+    [
+        pytest.param(ModelSummary.MODE_FULL),
+        pytest.param(ModelSummary.MODE_TOP),
+    ],
+)
```
```suggestion
@pytest.mark.parametrize(["mode"], ModelSummary.MODES)
```

carmocca commented on tests/core/test_memory.py:
```diff
@@ -186,36 +238,92 @@ def test_summary_parameter_count(mode):
+        pytest.param(ModelSummary.MODE_TOP),
+    ],
+)
+@pytest.mark.parametrize(
+    ["example_input", "expected_model_size"],
+    [
+        pytest.param(torch.zeros(1, 1, 28, 28), 0.318),
+        pytest.param(torch.zeros(1, 1, 224, 224), 31.84),
+        pytest.param(torch.zeros(10, 1, 512, 512), 183.425),
+        pytest.param(None, 0.075),
+    ],
+)
```
```suggestion
@pytest.mark.parametrize(
    ["example_input", "expected_model_size"],
    [
        (torch.zeros(1, 1, 28, 28), 0.318),
        (torch.zeros(1, 1, 224, 224), 31.84),
        (torch.zeros(10, 1, 512, 512), 183.425),
        (None, 0.075),
    ],
)
```

carmocca commented on tests/core/test_memory.py:
```diff
@@ -186,36 +238,92 @@ def test_summary_parameter_count(mode):
+    assert almost_equals(summary.model_size(), expected_model_size, rel_tol=1e-3, abs_tol=1e-3)
+
+
+@pytest.mark.parametrize(
+    ["mode"],
+    [
+        pytest.param(ModelSummary.MODE_FULL),
+    ],
+)
```
We don't need this parametrize if there is only one case, you can set the mode in the function

kartik4949 replied:
Yeah but just for consistency in test functions i did this
I can explicitly initialize summary with mode too :) should I?

carmocca replied:
Sure, you are only testing for that single case here.

kartik4949 replied:
ok

carmocca commented on tests/core/test_memory.py:
```diff
@@ -186,36 +238,92 @@ def test_summary_parameter_count(mode):
+@pytest.mark.parametrize(
+    ["example_input", "expected_model_size"],
+    [
+        pytest.param(torch.zeros(10, 256), 0.527),
+        pytest.param(None, 0.505),
```
```suggestion
        (torch.zeros(10, 256), 0.527),
        (None, 0.505),
```

carmocca commented on tests/core/test_memory.py:
```diff
@@ -186,36 +238,92 @@ def test_summary_parameter_count(mode):
+    assert almost_equals(summary.model_size(), expected_model_size, rel_tol=1e-3, abs_tol=1e-3)
+
+
+@pytest.mark.parametrize(
+    ["mode"],
+    [
+        pytest.param(ModelSummary.MODE_FULL),
+        pytest.param(ModelSummary.MODE_TOP),
+    ],
+)
```
```suggestion
@pytest.mark.parametrize(["mode"], ModelSummary.MODES)
```

carmocca commented on tests/core/test_memory.py:
```diff
@@ -186,36 +238,92 @@ def test_summary_parameter_count(mode):
+@pytest.mark.parametrize(
+    ["example_input", "expected_size"],
+    [
+        pytest.param([], UNKNOWN_SIZE),
+        pytest.param((1, 2, 3), [UNKNOWN_SIZE] * 3),
+        pytest.param(torch.tensor(0), UNKNOWN_SIZE),
+        pytest.param(dict(tensor=torch.zeros(1, 2, 3)), UNKNOWN_SIZE),
+        pytest.param(torch.zeros(2, 3, 4), [2, 3, 4]),
+        pytest.param([torch.zeros(2, 3), torch.zeros(4, 5)], [[2, 3], [4, 5]]),
+        pytest.param((torch.zeros(2, 3), torch.zeros(4, 5)), [[2, 3], [4, 5]]),
```
```suggestion
        ([], UNKNOWN_SIZE),
        ((1, 2, 3), [UNKNOWN_SIZE] * 3),
        (torch.tensor(0), UNKNOWN_SIZE),
        (dict(tensor=torch.zeros(1, 2, 3)), UNKNOWN_SIZE),
        (torch.zeros(2, 3, 4), [2, 3, 4]),
        ([torch.zeros(2, 3), torch.zeros(4, 5)], [[2, 3], [4, 5]]),
        ((torch.zeros(2, 3), torch.zeros(4, 5)), [[2, 3], [4, 5]]),
```

carmocca commented on tests/core/test_memory.py:
```diff
@@ -20,6 +20,43 @@
 from tests.base.models import ParityModuleRNN
 
 
+def almost_equals(a, b, rel_tol=0.0, abs_tol=0.0):
+    _almost_close = lambda a, b: abs(a - b) <= max(rel_tol * max(abs(a), abs(b)), abs_tol)
+    return _almost_close(a, b)
```
You can use `torch.allclose` or `torch.testing.assert_allclose` for this

kartik4949 replied:
oh nice will do it, im new to torch :P

kartik4949 commented on pytorch_lightning/core/memory.py:
```diff
@@ -213,6 +220,100 @@ def out_sizes(self) -> List:
+
+        NOTE: Currently only Supported in Full Mode.
+
+        ::
```
Only supported in Full Summary mode, model_size

carmocca replied:
I mean the `::`

kartik4949 replied:
minor error.
meant Example::

kartik4949 commented on pytorch_lightning/core/memory.py:
```diff
@@ -213,6 +220,100 @@ def out_sizes(self) -> List:
+                    # do nothing if tried to find prod on unknown type.
+                    pass
+            else:
+                _ = [_get_out_size_params(out_size) for out_size in out_sizes if isinstance(out_size, list)]
```
Makes sense :)

kartik4949 commented on pytorch_lightning/core/memory.py:
```diff
@@ -213,6 +220,100 @@ def out_sizes(self) -> List:
 
+    @property
+    def total_params(self) -> int:
+        _total_params = sum(p.numel() for p in self._model.parameters())
+        return _total_params
```
sure

kartik4949 commented on pytorch_lightning/core/memory.py:
```diff
@@ -213,6 +220,100 @@ def out_sizes(self) -> List:
+        ]
+
+        # recursive traversal to calculate output size.
+        # recursive is used to handle nested output sizes i.e [[[1,2,3],[[12,2,3], [1,3,4]]], [2,3,4]].
```
@carmocca whats the formatter u r using.

carmocca replied:
I use `black -l 120 -S` in this project

kartik4949 replied:
cool let met format it so

kartik4949 commented on pytorch_lightning/core/memory.py:
```diff
@@ -367,12 +485,14 @@ def get_gpu_memory_map() -> Dict[str, int]:
 
 
+def get_formatted_model_size(total_model_size: float) -> float:
+    return "{:.3f}".format(total_model_size)
```
@carmocca good catch f-strings are better.

kartik4949 commented on pytorch_lightning/core/memory.py:
```diff
@@ -213,6 +220,100 @@ def out_sizes(self) -> List:
+
+    def total_out_params(self, batch_size_dim: int) -> int:
+        """ finds total output parameters to calculate forward/backward pass size. """
+        _total_out_params = 0
+
+        _out_sizes = [
```
here because of name clashes :)

kartik4949 commented on pytorch_lightning/core/memory.py:
```diff
@@ -213,6 +220,100 @@ def out_sizes(self) -> List:
+            >> summary.model_size()
+
+        Returns:
+            float: Total estimated model size(MB) if example input array is passed else Total Model Params Size(MB).
```
auto docstring generator :P

awaelchli reviewed (changes requested):
can we aim cleaner implemenation that makes fewer assumptions about the model?
also how good is this estimate if most models are not fully modular?

kartik4949 commented:
@awaelchli it works with nested sequential and modular models.

kartik4949 commented:
@awaelchli only assumptions are for input type. 
And the complex code under def _total_out_params size 
Is to calculate forward and backward pass!
If we will consider only total params as model size then it will be a cleaner implementation 
But i suggest to show total model size with forward pass too :)

kartik4949 reviewed (commented):
> can we aim for a cleaner implemenation that makes fewer assumptions about the model?
> also how good is this estimate if most models are not fully modular?

Please see the nested_sequential test :)

tchaton commented:
Hey @kartik4949,

Any updates on tis PR ?

Best,
T.C

kartik4949 commented:
The implementation works with every mode i tested 
May be some refactors :) 
If anyone want to help :)

kartik4949 commented:
@awaelchli @justusschock  Hey
lets calculate model forward pass size only 
as it will have lesser complication and further we can add more feature to this one.
what you say ?
if so then I can remove the complex feature.
lets conclude this PR faster

justusschock commented:
@kartik4949 Yes, I think that's best. And for the more complicated part, maybe it's better to do this in a separate follow up PR :)

kartik4949 commented:
@justusschock thanks for responding will remove other sizes and conclude this one :)

kartik4949 commented:
@justusschock hey, done with changes.
btw i did rebase on 1.15 
but pr is for 1.12-dev.
idk how to interpret this :)

kartik4949 commented:
closing as pr is bit messy, starting fresh
